src/dvc_data/hashfile/checkout.py

```diff
@@ -144,24 +148,33 @@ def _needs_relink(
     return True
 
 
+def _change_needs_relink(
+    change: "Change", path: str, fs: "FileSystem", cache: "HashFileDB"
+) -> tuple["Change", bool]:
+    meta = change.old.meta
+    if meta is None:
+        return change, True
+
+    assert meta is not None
+    p = fs.sep.join((path, *change.old.key))
+    oid = change.new.oid.value if change.new.oid else None
+    relink = _needs_relink(p, cache, meta, oid)
+    return change, relink
+
+
 def _check_relink(
     diff: DiffResult, path: str, fs: "FileSystem", cache: "HashFileDB"
 ) -> None:
     if not isinstance(fs, LocalFileSystem) or not isinstance(cache.fs, LocalFileSystem):
         diff.modified.extend(diff.unchanged)
         return
 
-    for change in diff.unchanged:
-        meta = change.old.meta
-        if meta is None:
-            diff.modified.append(change)
-            continue
-
-        assert meta is not None
-        p = fs.sep.join((path, *change.old.key))
-        oid = change.new.oid.value if change.new.oid else None
-        if _needs_relink(p, cache, meta, oid):
-            diff.modified.append(change)
+    relink_func = partial(_change_needs_relink, fs=fs, path=path, cache=cache)
+    max_workers = min(16, (os.cpu_count() or 1) + 4)
```

Comment:
Taken from here but not sure why the thread pool should depend on cpu count:

https://github.com/iterative/dvc-data/blob/6ad5866b56b972128f90fd20e4d297ac45fe59fa/src/dvc_data/hashfile/build.py#L99